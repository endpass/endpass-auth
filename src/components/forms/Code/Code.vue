<template>
  <component
    :is="currentForm"
    v-bind="$attrs"
    v-on="$listeners"
  />
</template>

<script>
import EmailCode from './EmailCode';
import OtpCode from './OtpCode';

export default {
  name: 'CodeForm',

  props: {
    isOtp: {
      type: Boolean,
      required: true,
    },
  },

  computed: {
    currentForm() {
      return this.isOtp ? OtpCode : EmailCode;
    },
  },

  components: {
    EmailCode,
    OtpCode,
  },
};
</script>
