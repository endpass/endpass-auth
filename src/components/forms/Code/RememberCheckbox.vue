<template>
  <v-checkbox
    v-bind="$attrs"
    type="checkbox"
    v-on="$listeners"
  >
    {{ $t('components.rememberCheckbox.text') }}
  </v-checkbox>
</template>

<script>
import VCheckbox from '@endpass/ui/kit/VCheckbox';

export default {
  name: 'RememberCheckbox',

  components: {
    VCheckbox,
  },

  model: {
    prop: 'value',
    event: 'change',
  },
};
</script>
