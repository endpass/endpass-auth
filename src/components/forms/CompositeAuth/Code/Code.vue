<template>
  <component
    :is="currentForm"
    :email="email"
    :password="password"
    @submit="onSubmit"
    @cancel="onCancel"
  />
</template>

<script>
import EmailCode from './EmailCode';
import OtpCode from './OtpCode';
import { authStore } from '@/store';

export default {
  name: 'Code',

  props: {
    email: {
      type: String,
      required: true,
    },

    password: {
      type: String,
      required: true,
    },
  },

  authStore,

  computed: {
    currentForm() {
      return this.otpEmail ? OtpCode : EmailCode;
    },

    otpEmail() {
      return this.$options.authStore.otpEmail;
    },
  },

  methods: {
    onCancel() {
      this.$emit('cancel');
    },

    onSubmit(code) {
      this.$emit('submit', code);
    },
  },

  components: {
    EmailCode,
    OtpCode,
  },
};
</script>
