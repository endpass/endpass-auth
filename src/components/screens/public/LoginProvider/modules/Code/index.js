export { default } from './Code.interface';
