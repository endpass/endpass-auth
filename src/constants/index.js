export const METHODS = {
  SIGN: 'SIGN',
  ACCOUNT: 'ACCOUNT',
  RECOVER: 'RECOVER',
  GET_SETTINGS: 'GET_SETTINGS',

  AUTH: 'AUTH',
  LOGOUT: 'LOGOUT',
  DIALOG_RESIZE: 'DIALOG_RESIZE',
  DIALOG_OPEN: 'DIALOG_OPEN',
  DIALOG_CLOSE: 'DIALOG_CLOSE',
  INITIATE: 'INITIATE',
  READY_STATE_BRIDGE: 'READY_STATE_BRIDGE',
};

export const LS_SETTINGS = 'endpass-connect:settings';

export const IDENTITY_MODE = Object.freeze({
  DEFAULT: 'default',
  CUSTOM: 'custom',
  LOCAL: 'local',
});

export const DIRECTION = Object.freeze({
  AUTH: 'auth',
  CONNECT: 'connect',
});

export const ORIGIN_HOST = (() => {
  const parser = document.createElement('a');
  parser.href = document.referrer;
  return parser.origin;
})();
