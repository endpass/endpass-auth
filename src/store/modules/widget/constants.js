export const WIDGET_RESIZE_DURATION = 300;

export default {};
