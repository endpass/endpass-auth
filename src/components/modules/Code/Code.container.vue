<template>
  <component
    :is="current"
    :is-loading="isLoading"
    :error="error"
    :email="email"
    @recover="onRecover"
    @send-code="onSendCode"
    @submit="onSubmit"
  />
</template>
<script>
import AppCode from './AppCode';
import EmailCode from './EmailCode';
import SmsCode from './SmsCode';
import { CHALLENGE_TYPES } from '@/constants';

const COMPONENT_BY_TYPE = {
  [CHALLENGE_TYPES.EMAIL_OTP]: EmailCode,
  [CHALLENGE_TYPES.APP_OTP]: AppCode,
  [CHALLENGE_TYPES.SMS_OTP]: SmsCode,
};

export default {
  name: 'CodeContainer',

  props: {
    isLoading: {
      type: Boolean,
      default: false,
    },

    error: {
      type: String,
      default: '',
    },

    challengeType: {
      type: String,
      required: true,
    },

    email: {
      type: String,
      required: true,
    },
  },

  computed: {
    current() {
      return COMPONENT_BY_TYPE[this.challengeType];
    },
  },

  methods: {
    onSendCode() {
      this.$emit('send-code');
    },

    onRecover() {
      this.$emit('recover');
    },

    onSubmit({ code }) {
      this.$emit('submit', { code });
    },

    onCancel() {
      this.$emit('cancel');
    },
  },

  components: {
    SmsCode,
    EmailCode,
    AppCode,
  },
};
</script>
