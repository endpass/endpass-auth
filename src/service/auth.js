// @ts-check
import get from 'lodash/get';
import request from '@/class/singleton/request';
import { AUTH_STATUS_CODE } from '@/constants';

const identityBaseUrl = ENV.VUE_APP_IDENTITY_API_URL;

const TIMEOUT_DEFAULT = 1500;

/**
 * @type {object}
 */
const CODE_TO_STATUS = {
  200: AUTH_STATUS_CODE.LOGGED_IN,
  401: AUTH_STATUS_CODE.NOT_LOGGED,
  403: AUTH_STATUS_CODE.NEED_PERMISSION,
};

/**
 * @param {Promise} promise
 * @return {Promise<any>}
 */
const withSuccess = promise => {
  return promise.then(res => {
    if (res && !res.success) throw new Error(res.message);

    return res;
  });
};

/**
 * @param {Function} handler
 * @return {number}
 */
const createTimeout = handler => setTimeout(handler, TIMEOUT_DEFAULT);

/**
 * @param {string} email
 * @return {Promise<any>}
 */
const getAuthChallenge = email => {
  return withSuccess(
    request.post(`${identityBaseUrl}/auth`, {
      email,
    }),
  );
};

/**
 * @param {string} password
 * @param {string} originHost
 * @return {Promise<any>}
 */
const setAuthPermission = async (password, originHost) => {
  const res = await request.post(`${identityBaseUrl}/auth/permission`, {
    password,
    originHost,
  });
  return res;
};

/**
 * @param {object} params
 * @param {string} params.email
 * @param {string} params.code
 * @param {string} params.password
 * @param {boolean} params.isSignUp
 * @param {boolean} params.isRemember
 * @return {Promise<any>}
 */
const authWithCode = async ({
  email,
  code,
  password,
  isSignUp,
  isRemember,
}) => {
  const url = isSignUp
    ? `${identityBaseUrl}/auth/signup`
    : `${identityBaseUrl}/auth/token`;

  return withSuccess(
    request.post(url, {
      email,
      code,
      password,
      remember: isRemember,
    }),
  );
};

/**
 * @param {string} idToken
 * @return {Promise<any>}
 */
const authWithGoogle = idToken =>
  request
    .post(`${identityBaseUrl}/auth/google`, {
      token: idToken,
    })
    .then(res => {
      if (!res.success) throw new Error(res.message);
      return res;
    })
    .catch(err => {
      throw err.response.data;
    });

/**
 * @param {string} code
 * @return {Promise<any>}
 */
const authWithGitHub = code =>
  request
    .post(`${identityBaseUrl}/auth/github`, {
      code,
    })
    .then(res => {
      if (!res.success) throw new Error(res.message);
      return res;
    })
    .catch(err => {
      throw err.response.data;
    });

const logout = () => request.post(`${identityBaseUrl}/logout`);

/**
 * @return {Promise<{expiresAt: number, status: string, hash: string}>}
 */
const getAuthStatus = async () => {
  try {
    const { expiresAt, hash } = await request.get(
      `${identityBaseUrl}/auth/check`,
    );
    return {
      status: AUTH_STATUS_CODE.LOGGED_IN,
      hash,
      expiresAt,
    };
  } catch (e) {
    const statusCode = get(e, ['response', 'status']);
    const status = CODE_TO_STATUS[statusCode] || AUTH_STATUS_CODE.NOT_LOGGED;

    const hash = get(e, ['response', 'data', 'hash'], '');
    const expiresAt = get(e, ['response', 'data', 'expiresAt'], 0);
    return {
      status,
      hash,
      expiresAt,
    };
  }
};

/**
 * @return {Promise<number>}
 */
const waitLogin = () =>
  new Promise((resolve, reject) => {
    /* eslint-disable-next-line */
    const handler = async function() {
      try {
        const { status } = await getAuthStatus();

        if (
          status === AUTH_STATUS_CODE.LOGGED_IN ||
          status === AUTH_STATUS_CODE.NEED_PERMISSION
        ) {
          return resolve();
        }

        createTimeout(handler);
      } catch (err) {
        return reject(err);
      }
    };

    return handler();
  });

/**
 * @param {string} email
 * @return {Promise<string>}
 */
const getRecoveryIdentifier = email =>
  request
    .get(`${identityBaseUrl}/auth/recover?email=${encodeURIComponent(email)}`)
    .then(res => {
      if (!res.success) throw new Error(res.message);

      return res.message;
    });

/**
 * @param {string} email
 * @return {Promise<number>}
 */
const sendEmailCode = async email => {
  const { timeout } = await request.post(`${identityBaseUrl}/auth/code`, {
    email,
  });

  return timeout;
};

/**
 * Send sms with code for disabling otp
 * @param {string} email
 * @returns {Promise<any>}
 */
const sendOtpRecoverSms = async email => {
  try {
    return withSuccess(
      request.get(
        `${identityBaseUrl}/auth/recover?email=${encodeURIComponent(email)}`,
      ),
    );
  } catch (error) {
    const { response = {} } = error;
    const { status } = response;
    error.code = status;

    throw error;
  }
};

/**
 * Disable otp setting
 * @param {object} param
 * @param {string} param.email
 * @param {string} param.code
 * @returns {Promise<any>}
 */
const disableOtpViaSms = async ({ email, code }) => {
  return withSuccess(
    request.post(`${identityBaseUrl}/auth/recover`, {
      email,
      code,
    }),
  );
};

/**
 * @return {string}
 */
const getSeedTemplateUrl = () => `${identityBaseUrl}/auth/recover/template`;

export default {
  getAuthStatus,
  setAuthPermission,
  getAuthChallenge,
  authWithCode,
  authWithGoogle,
  authWithGitHub,
  sendEmailCode,
  logout,
  waitLogin,
  getRecoveryIdentifier,
  sendOtpRecoverSms,
  disableOtpViaSms,
  getSeedTemplateUrl,
};
