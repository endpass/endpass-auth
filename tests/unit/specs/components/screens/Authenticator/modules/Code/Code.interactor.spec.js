import Vuex from 'vuex';
import { shallowMount, createLocalVue } from '@vue/test-utils';
import { email, regularPassword as password, code } from '@unitFixtures/auth';
import CodeInteractor from '@/components/screens/Authenticator/modules/Code/Code.interactor';
import Code from '@/components/modules/Code';
import setupI18n from '@/locales/i18nSetup';
import authService from '@/service/auth';
import { CHALLENGE_TYPES } from '@/constants';

const localVue = createLocalVue();
localVue.use(Vuex);
const i18n = setupI18n(localVue);

describe('CodeInteractor', () => {
  let wrapper;
  const challengeType = CHALLENGE_TYPES.EMAIL_OTP;

  const createWrapper = (options, props) =>
    shallowMount(CodeInteractor, {
      localVue,
      propsData: {
        email,
        password,
        challengeType,
        isSignUp: false,
        ...props,
      },
      i18n,
      ...options,
    });

  beforeEach(async () => {
    jest.clearAllMocks();

    wrapper = createWrapper();
    await global.flushPromises();
  });

  describe('render', () => {
    it('should correctly render component', () => {
      expect(wrapper.name()).toBe('CodeInteractor');
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('should render form', () => {
      expect(wrapper.find('code-stub').exists()).toBe(true);
      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe('behavior', () => {
    describe('submit', () => {
      const defaultServiceParams = {
        code,
        password,
        email,
        isSignUp: false,
      };

      it('should handle submit event', async () => {
        expect.assertions(3);

        expect(authService.authWithCode).not.toBeCalled();

        wrapper.find(Code).vm.$emit('submit', { code });

        expect(authService.authWithCode).toBeCalledTimes(1);
        expect(authService.authWithCode).toBeCalledWith({
          ...defaultServiceParams,
        });
      });

      it('should handle submit event with sign up prop', async () => {
        expect.assertions(3);

        wrapper = createWrapper({}, { isSignUp: true });
        await global.flushPromises();

        expect(authService.authWithCode).not.toBeCalled();

        wrapper.find(Code).vm.$emit('submit', { code });

        expect(authService.authWithCode).toBeCalledTimes(1);
        expect(authService.authWithCode).toBeCalledWith({
          ...defaultServiceParams,
          isSignUp: true,
        });
      });

      it('should handle submit event with remember prop', async () => {
        expect.assertions(3);

        expect(authService.authWithCode).not.toBeCalled();

        wrapper.find(Code).vm.$emit('submit', { code, isRemember: true });

        expect(authService.authWithCode).toBeCalledTimes(1);
        expect(authService.authWithCode).toBeCalledWith({
          ...defaultServiceParams,
          isRemember: true,
        });
      });

      it('should not handle submit event when loading status true', async () => {
        expect.assertions(2);

        expect(authService.authWithCode).not.toBeCalled();

        wrapper.find(Code).vm.$emit('submit', { code });
        wrapper.find(Code).vm.$emit('submit', { code });

        expect(authService.authWithCode).toBeCalledTimes(1);
      });

      describe('loading status', () => {
        it('should be false before submit', () => {
          expect(wrapper.find('code-stub').attributes().isloading).toBeFalsy();
        });

        it('should be true after submit', () => {
          wrapper.find(Code).vm.$emit('submit', { code });

          expect(wrapper.find('code-stub').attributes().isloading).toBe('true');
        });

        it('should be false after handling submit', async () => {
          expect.assertions(1);

          wrapper.find(Code).vm.$emit('submit', { code });
          await global.flushPromises();

          expect(wrapper.find('code-stub').attributes().isloading).toBeFalsy();
        });
      });

      describe('error', () => {
        beforeEach(() => {
          authService.authWithCode.mockRejectedValueOnce(new Error('error'));
        });

        it('should pass error', async () => {
          expect.assertions(2);

          expect(wrapper.find('code-stub').attributes().error).toBeFalsy();

          wrapper.find(Code).vm.$emit('submit', { code });
          await global.flushPromises();

          expect(wrapper.find('code-stub').attributes().error).toBe(
            i18n.t('components.code.authFailed'),
          );
        });

        it('should remove error if exists before submit', async () => {
          expect.assertions(1);

          wrapper.find(Code).vm.$emit('submit', { code });
          await global.flushPromises();

          wrapper.find(Code).vm.$emit('submit', { code });

          expect(wrapper.find('code-stub').attributes().error).toBeFalsy();
        });
      });
    });

    describe('recover', () => {
      it('should emit recover event', () => {
        expect(wrapper.emitted().recover).toBeUndefined();

        wrapper.find(Code).vm.$emit('recover');

        expect(wrapper.emitted().recover.length).toBe(1);
        expect(wrapper.emitted().recover[0]).toEqual([]);
      });

      it('should not emit recover event while loading', () => {
        wrapper.find(Code).vm.$emit('submit', { code });
        wrapper.find(Code).vm.$emit('recover');

        expect(wrapper.emitted().recover).toBeUndefined();
      });
    });
  });
});
