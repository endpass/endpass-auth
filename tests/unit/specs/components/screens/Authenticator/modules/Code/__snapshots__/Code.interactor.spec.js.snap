// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodeInteractor render should correctly render component 1`] = `<code-stub error="" challengetype="EMAIL_OTP" email="user@email.com"></code-stub>`;

exports[`CodeInteractor render should render form 1`] = `<code-stub error="" challengetype="EMAIL_OTP" email="user@email.com"></code-stub>`;
